### PR TITLE
IBP-4084 Import Germplasm Updates

### PR DESCRIPTION
**Pull request by abatac**

### CODE REVIEW

**Review by clarysabel — COMMENTED** [review]
Hi @abatac I could not review all the code yet but I am leaving some comments, I am concerned about how you handle the case where the same code is an attribute and a name type 
I will continue tomorrow 
Claudia could not complete testing my ticket today but my review is already done and approved by Nahuel 
CC @darla-leafnode

File: src/main/java/org/generationcp/middleware/api/germplasm/update/GermplasmUpdateService.java
Comment (abatac) [reply src/main/java/org/generationcp/middleware/api/germplasm/update/GermplasmUpdateService.java:7]: I agree, we can have one service for import and update. Will adjust my code after your branch is merged.

File: src/main/java/org/generationcp/middleware/domain/germplasm/GermplasmUpdateDTO.java
Comment (abatac) [reply src/main/java/org/generationcp/middleware/domain/germplasm/GermplasmUpdateDTO.java:22]: I also thought about separating the names from attributes. But that means the codes must be parsed and separated at the client level. I just find it convenient to detect which codes are names and attributes in the backend, reducing complexity in determining what the code is in the client.

During validation (org.ibp.api.java.impl.middleware.common.validator.GermplasmUpdateValidator#validateCodes), the service checks if a code is defined as both name and attribute and throw a validation error before the update process proceeds.

File: src/main/java/org/generationcp/middleware/service/api/userdefinedfield/UserDefinedFieldService.java
Comment (abatac) [reply src/main/java/org/generationcp/middleware/service/api/userdefinedfield/UserDefinedFieldService.java]: Done, I reused the existing method org.generationcp.middleware.manager.api.GermplasmDataManager#getUserDefinedFieldByTableTypeAndCodes.

File: src/main/resources/mw_messages_en.properties
Comment (abatac) [reply src/main/resources/mw_messages_en.properties:23]: This is handled in org.ibp.api.java.impl.middleware.common.validator.GermplasmUpdateValidator#validateCodes.

File: src/main/java/org/generationcp/middleware/dao/GermplasmDAO.java
Comment (abatac) [reply src/main/java/org/generationcp/middleware/dao/GermplasmDAO.java]: I already removed this.

File: src/main/java/org/generationcp/middleware/api/germplasm/update/GermplasmUpdateService.java
Comment (abatac) [reply src/main/java/org/generationcp/middleware/api/germplasm/update/GermplasmUpdateService.java]: This is done. I deleted the GermplasmUpdateService and moved the logic to GermplasmService. 

I didn't reuse the GermplasmImportRequestDto.GermplasmDto for importing germplasm updates since it requires gid and creation date must be in yyyyMMdd format.

File: src/main/java/org/generationcp/middleware/manager/api/GermplasmDataManager.java
Comment (abatac) [reply src/main/java/org/generationcp/middleware/manager/api/GermplasmDataManager.java]: Done.

File: src/main/java/org/generationcp/middleware/api/breedingmethod/BreedingMethodService.java
Comment (abatac) [reply src/main/java/org/generationcp/middleware/api/breedingmethod/BreedingMethodService.java]: Done.

File: src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java
Comment (clarysabel) [inline src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java:299]: Is this format needed? I saw it when testing and it shows: 99,641 and the , is kind of weird. I think we can just convert the int to String without any format

File: src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java
Comment (clarysabel) [inline src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java:312]: There is a piece of the requirement that it is not implemented yet, gnpgs, gpid1 and gpid2 may be updated depending on the method change. I knew from Mariano that you could not make it for the sprint but I strongly suggest to include it and merge this one when it is included no matter if we need to wait to merge it next sprint.

File: src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java
Comment (clarysabel) [inline src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java:379]: I am not sure If I am taking care of this case on the import, I will check

**Review by clarysabel — CHANGES_REQUESTED** [review]
Leaving some comments

File: src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java
Comment (abatac) [reply src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java:136]: In my testing and QA branch testing, attributes are added if not exist.

Comment (abatac) [reply src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java:136]: This is indeed a bug. Updating and creating new attributes only worked for ATRIBUTS' ATTRIBUTE type. I just had to include ATRIBUTS' PASSPORT type in retrieving attributes. This is now fixed.

File: src/main/java/org/generationcp/middleware/domain/germplasm/GermplasmUpdateDTO.java
Comment (abatac) [reply src/main/java/org/generationcp/middleware/domain/germplasm/GermplasmUpdateDTO.java:16]: I tried the JsonFormat approach but it's now throwing an error if the date string is invalid. So I decided to validate it through our code.

File: src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java
Comment (abatac) [reply src/main/java/org/generationcp/middleware/api/germplasm/GermplasmServiceImpl.java:302]: I can't replicate the issue you mentioned. Have you tried to recompile? The string format is needed to combine the method name and the type so that the message will show like this:

_Cannot change method for germplasm **1**, the breeding method should be the same type as **Unknown bulking method (DER**)_

File: src/main/java/org/generationcp/middleware/domain/germplasm/GermplasmUpdateDTO.java
Comment (clarysabel) [reply src/main/java/org/generationcp/middleware/domain/germplasm/GermplasmUpdateDTO.java:16]: Hi @abatac I changed my data type to String as well. I had some conflicts with spring boot version while upgrading jackson so it was easier for now to change it from Date to String

**Pull request merged**
